torchaudio/csrc/build_rir.cpp

```diff
@@ -0,0 +1,134 @@
+#include <math.h>
+#include <torch/script.h>
+#include <torch/torch.h>
+using namespace torch::indexing;
+
+namespace torchaudio {
+namespace rir {
+
+template <typename scalar_t>
+void build_rir_impl(
+    const torch::Tensor& irs,
+    const torch::Tensor& delay,
+    torch::Tensor& rirs,
+    const int64_t rir_length,
+    const int64_t num_band,
+    const int64_t num_image,
+    const int64_t num_mic,
+    const int64_t ir_length) {
+  const scalar_t* input_data = irs.data_ptr<scalar_t>();
+  const int* delay_data = delay.data_ptr<int>();
+  scalar_t* output_data = rirs.data_ptr<scalar_t>();
+  at::parallel_for(
+      0, num_band * num_image * num_mic, 0, [&](int64_t start, int64_t end) {
+        for (auto i = start; i < end; i++) {
+          int64_t offset_input = i * ir_length;
+          int64_t mic = i % num_mic;
+          int64_t image = ((i - mic) / num_mic) % num_image;
+          int64_t band = (i - mic - image * num_mic) / (num_image * num_mic);
+          int64_t offset_output = (band * num_mic + mic) * rir_length;
+          int64_t offset_delay = image * num_mic + mic;
+          for (auto j = 0; j < ir_length; j++) {
+            output_data[offset_output + j + delay_data[offset_delay]] +=
+                input_data[offset_input + j];
+          }
+        }
+      });
+}
+
+torch::Tensor build_rir(
+    const torch::Tensor irs,
+    const torch::Tensor delay,
+    const int64_t rir_length) {
+  const int64_t num_band = irs.size(0);
+  const int64_t num_image = irs.size(1);
+  const int64_t num_mic = irs.size(2);
+  const int64_t ir_length = irs.size(3);
+  torch::Tensor rirs =
+      torch::zeros({num_band, num_mic, rir_length}, irs.dtype());
+  rirs.requires_grad_(true);
```

Comment:
Does this work with `torch.inference_mode()`?

Reply:
actually, what does it accomplish?
`build_rir_impl` manipulates the data of `rirs` by pointer, so autograd cannot track it.
In this case, the modification of `requires_grad=True` should be done by client code.

Reply:
I tried to remove this line and run autograd_test, the test fails but was working with this line.

Reply:
It also passes the autograd test.